modules/packages/Channel.chpl

```diff
@@ -0,0 +1,605 @@
+        type eltType;
+
+        /* Maximum number of elements that the channel can hold at a time. */
+        var bufferSize : int;
+        var buffer : [0..#bufferSize] eltType;
+        var sendIdx = 0;
+        var recvIdx = 0;
+        var count = 0;
+        var closed = false;
+        var sendWaiters : WaiterQueue;
+        var recvWaiters : WaiterQueue;
+        var lock$ = new _LockWrapper();
+
+        proc init(type elt, size = 0) {
+            eltType = elt;
+            bufferSize = size;
+            sendWaiters = new WaiterQueue(elt);
+            recvWaiters = new WaiterQueue(elt);
+        }
+
+        proc lock() {
+            lock$.lock();
+        }
+
+        proc unlock() {
+            lock$.unlock();
+        }
+
+        proc recv(out val : eltType, blocking : bool) : bool {
+            if blocking then lock();
+
+            if closed && count == 0 {
+                if blocking then unlock();
+                if !blocking then return true;
+                else return false;
+            }
+
+            while !sendWaiters.isEmpty() && sendWaiters.front!.isSelect {
+                if sendWaiters.front!.isSelectDone.deref().compareAndSwap(-1, sendWaiters.front!.selectId) {
+                    break;
+                }
+                else sendWaiters.dequeue();
+            }
+
+            if count == 0 && sendWaiters.isEmpty() {
+                if !blocking then return false;
+                var process$ : single bool;
+                var processing = new unmanaged Waiter(val, process$);
+                recvWaiters.enqueue(processing);
+                unlock();
+                var status = processing.suspend();
+                delete processing;
+                return status;
+            }
+
+            if bufferSize > 0 {
+                val = buffer[recvIdx];
+            }
+
+            if !sendWaiters.isEmpty() {
+                var sender = sendWaiters.dequeue();
+                if bufferSize > 0 {
+                    buffer[recvIdx] = sender.val.deref();
+
+                    sendIdx = (sendIdx + 1) % bufferSize;
+                    recvIdx = (recvIdx + 1) % bufferSize;
+                }
+                else val = sender.val.deref();
+
+                sender.notify(true);
+            }
+            else {
+
+                recvIdx = (recvIdx + 1) % bufferSize;
+                count -= 1;
+
+            }
+            if blocking then unlock();
+
+            return true;
+
+        }
+
+        proc send(in val : eltType, blocking : bool) : bool throws {
+            if blocking then lock();
+
+            if closed {
+                if blocking then unlock();
+                throw new owned ChannelError("Trying to send on a closed channel");
+            }
+
+            while !recvWaiters.isEmpty() && recvWaiters.front!.isSelect {
+                if recvWaiters.front!.isSelectDone.deref().compareAndSwap(-1, recvWaiters.front!.selectId) {
```

Comment:
Similarly, I imagine a helper here can help make this line `if recvWaiters.front!.update() {` or something